2단계 - 문자열 계산기

yeonsuyang opened the pull request:
안녕하세요!
2단계 문자열 계산기 구현하였습니다.
잘부탁드립니다.
감사합니다~

kakao-celine-kk committed:
feat. 2단계 - 문자열 계산기

kakao-celine-kk committed:
change. size exception 문구

laco-dev commented on src/test/kotlin/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,53 @@
+class CalculatorTest {
+
+    @ParameterizedTest
+    @CsvSource("2 + 3 * 4 / 2,10", "2 + 2,4", "4 - 1,3", "1 * 2,2", "4 / 1,4")
```
테스트의 기댓값과 결과를 명확하게 판단하기 위해서, 조금의 개행 조미료를 뿌리면 조금 더 가독성이 올라갈 수 있어요.
```kotlin
@CsvSource(
    "2 + 3 * 4 / 2 = 10",
    "2 + 2 = 4",
    "4 - 1 = 3",
    "1 * 2 = 2",
    "4 / 1 = 4",
    delimiter = '='
)
```

laco-dev commented on src/test/kotlin/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @Test
+    internal fun `입력값이 null이거나 빈 공백 문자열일 경우`() {
```
null 이거나 빈 공백 문자일 경우 --> `성공한다?` `실패한다?`
이 테스트가 실패한 경우, 콘솔에서 왜 실패했는지 파악하기 쉽도록 결과를 써보면 어떨까요?

laco-dev commented on src/test/kotlin/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,53 @@
+            Calculator.inputString(" ")
+        }
+
+        assertThatIllegalArgumentException().isThrownBy {
+            Calculator.inputString(null)
+        }
```
코틀린에서는 보다 간단한 사용 방법의 throws를 제공합니다.
아래 모두 동일합니다 🙏🏻 
```
assertThrows<IllegalArgumentException> { ... }
```

laco-dev commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,29 @@
+package calculator
+
+import java.lang.IllegalArgumentException
```
코틀린 패키지의 IlleaglArgumentException을 사용해보면 어떨까요?

코틀린에서는 이 방법을 사용하는 편 : https://github.com/next-step/kotlin-racingcar/pull/1032#discussion_r1221725373

laco-dev commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,29 @@
+
+private const val ZERO_EXCEPTION = "0으로는 나눌 수 없습니다."
+
+enum class Operator : BinaryOperator<Int>, IntBinaryOperator {
```
함수형 인터페이스를 이용해, Operator라는 객체가 연산을 수행하도록 잘 작성 해 주셨습니다.
보다 코틀린스럽게 만들어 본다면 고차함수를 이용해 볼 수 있습니다.
```kotlin
enum class Operator(
    private val operate: (Int, Int) -> Int,
) {
    PLUS({ a, b -> a + b }),
    ...
}
```

laco-dev commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+private const val SIZE_EXCEPTION = "계산을 위한 입력이 충분하지 않습니다."
+
+class Calculator {
+    companion object {
```
클래스 인스턴스로 사용되지 않는다면 코틀린에서는 object로 선언을 할 수 있습니다.

laco-dev commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+            return result
+        }
+
+        private fun calculate(operator: String, left: Int, right: Int): Int {
+            return when (operator) {
+                "+" -> Operator.PLUS.applyAsInt(left, right)
+                "-" -> Operator.MINUS.applyAsInt(left, right)
+                "*" -> Operator.MUL.applyAsInt(left, right)
+                "/" -> Operator.DIV.applyAsInt(left, right)
+                else -> throw IllegalArgumentException(OPERATOR_EXCEPTION)
+            }
+        }
```
Operator가 어떤 symbol을 가지고 있는지를 Calculator가 알아야 할 필요가 있을까요?
Calculator 입장에서는 어떤 Operator인지 관심 없이 연산을 하라는 협력으로의 관계만 유지할 수 있습니다.
아래 스닙핏을 참고 해보세요

```kotlin
fun calculate(...): Int {
   val operator: Operator = Operator.from(...)
   return operator.operate(left, right)
}
```

laco-dev commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+        fun inputString(input: String?): Int {
+            if (input.isNullOrBlank()) throw IllegalArgumentException(BLANK_EXCEPTION)
+            val inputArray = input.split(DELIMITER)
+            if (inputArray.size < 3) throw IllegalArgumentException(SIZE_EXCEPTION)
```
IlleaglArgumentException을 처리할 때에는 코틀린에서는 require, check 를 사용할 수 있습니다

laco-dev commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+
+class Calculator {
+    companion object {
+        fun inputString(input: String?): Int {
```
계산을 하는 협력이라는 이름으로 inputString(...) 이라고 작성되는게 맞을까요?

계산기에게 inputString 해줘! 라고 호출하는 것 같지 않나요?

laco-dev reviewed (changes requested):
계산기 미션 빠르게 진행 해 주셨네요. 👍🏻 
객체지향 학습과 더불어 보다 코틀린스럽게 작성하기 위한 연습을 앞으로 같이 해 나가면 될 것 같아요

본격적으로 자동차 미션 진행하기 앞서 고민해볼만한 내용을 몇가지 남겨드렸으니 반영 후 다시 리뷰 요청 주세요. 🙏🏻

kakao-celine-kk committed:
refactor: 테스트 코드 가독성 향상

kakao-celine-kk committed:
refactor: 계산기 코드 수정
1. 메서드 이름 수정
2. operator 관련코드 전부 operator로 이동
3. 고차함수 적용
4. class -> object 로 변경

yeonsuyang commented:
수정사항 반영하였습니다! 감사합니다

laco-dev commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,23 @@
+private const val SIZE_EXCEPTION = "계산을 위한 입력이 충분하지 않습니다."
+
+object Calculator {
+    fun getInput(input: String?): Int {
```
Input의 늪에서 벗어나지 못하는군요!
계산기에게 Input을 돌려줘가 맞을지 100번 고민 해 보시면 됩니다!

함수 시그니처가 아래와 같다면?
```
fun getInput(input: String?): Input { ... }
```

yeonsuyang replied:
안녕하세요. 
계산기의 기능에 대해 생각했을 때
1. 입력을 받는다.
2. 계산한다.
2가지의 역할이 있다고 생각하였습니다.

main을 따로 나누었다면 calculator의 역할은 calculate밖에없었겠지만, 메인이 따로 없다면 입력을 받아야 계산이 이루어지는것이기에 해당 네이밍에 큰 문제가 없다고 생각하였습니다. 혹은 execute, calculate의 중복사용과 같은 네이밍밖에 떠오르지않았고 가장 직관적인 것을 생각하였습니다.
 "함수 시그니처가 아래와 같다면" 으로 설명해주신 부분이 잘 이해가 안갑니다ㅠㅠ
혹 저렇게 된다면, 다른 이름을 가져야된다고 생각하시는걸까요?

laco-dev replied:
@yeonsuyang 

1. Calculator라는 객체가 메인에서 사용되는지, 콘솔에서 사용되는지 관심을 가질 필요가 있을까요?
2. getInput 이라는 getter() 이름으로 사용이 됩니다. 그렇다면 반환 타입도 Input을 반환할 것 처럼 느껴집니다. 그렇다면, Input을 반환하는 것이 계산 결과이다 라는 것을 사용하는 쪽에서 알 수 있을까요?

laco-dev reviewed (approved):
피드백 반영 잘 해주셨습니다.

각 피드백을 반영할 때에는 리뷰 한개당 하나의 커밋으로 만들어서 리뷰에 같이 남겨주시면 컨텍스트 파악에 많은 도움이 됩니다!
(실무에서도 좋은 기술입니다)

간단한 의견 하나 남겨두었는데, 참고 해 주시면 좋겠습니다.
다음 자동차 경주 미션 진행 해주세요. 🙏🏻
